### PR TITLE
Replace `private` modifier node with field on supporting types

### PR DESCRIPTION
**Pull request by kazcw**
### Pull Request Description

Eliminate `private` modifier-node: `private` is a field in supporting types, or a single-token node in the case of `private` declarations.

### Important Notes

- Rust parser tests: Switch to a builder-style API for defining expected `Function` ASTs to allow further changes to `Function` fields without rewriting all the tests again.
- TreeToIr: Fix discarded module-level `diag`; add a test that covers module diagnostics.
- Syntax: Disallow `private` methods in function blocks. (Previously this was enforced in the compiler.)

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [ ] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [ ] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [ ] Unit tests have been written where possible.
- [ ] If meaningful changes were made to logic or tests affecting Enso Cloud integration in the libraries, 
      or the Snowflake database integration, a run of the [Extra Tests](https://github.com/enso-org/enso/actions/workflows/extra-nightly-tests.yml) has been scheduled.
  - If applicable, it is suggested to paste a link to a successful run of the Extra Tests.

### CODE REVIEW

**Review by Akirathan — APPROVED** [review]

File: engine/runtime-integration-tests/src/test/java/org/enso/compiler/test/ErrorCompilerTest.java [hunk 1]
```diff
@@ -485,14 +485,24 @@ public void illegalPrivateKeywordUseInType() throws Exception {
         18);
   }
 
+  @Test
+  public void illegalPrivateKeywordRepeatedDeclarations() throws Exception {
+    var ir = parse("""
+        private
+        private
+        """);
+    assertSingleSyntaxError(
```
Comment (JaroslavTulach) [inline engine/runtime-integration-tests/src/test/java/org/enso/compiler/test/ErrorCompilerTest.java:494]: Can we (should we) verify:
```ruby
private

private plus x y = x + y
```
works? CCing @AdRiley.

**Review by JaroslavTulach — APPROVED** [review]

File: lib/rust/parser/src/syntax/statement.rs [hunk 2]
```diff
@@ -201,25 +173,76 @@ fn parse_statement<'s>(
...
+        let keyword = into_private_keyword(token);
+        let private = Tree::private(keyword);
+        statement = match statement.take() {
+            Some(statement) => Tree::app(
+                private.with_error(match visibility_context {
+                    VisibilityContext::Public => SyntaxError::StmtUnexpectedPrivateSubject,
+                    VisibilityContext::Private => SyntaxError::StmtUnexpectedPrivateContext,
+                }),
+                statement,
+            ),
```
Comment (farmaazon) [inline lib/rust/parser/src/syntax/statement.rs:199]: It took me a good while to understand, why do we always set error here. I assumed that the responsibility of this function is to make statement a private object, not that it is already applied and any _excessive_ private is an error. Please rename/add short docs for this function.

File: lib/rust/parser/src/syntax/statement/function_def.rs [hunk 3]
```diff
@@ -87,7 +88,15 @@ pub fn parse_constructor_definition<'s>(
...
-    Tree::constructor_definition(name, inline_args, block_args)
+    let private =
+        (private_keywords_start < start).then(|| into_private_keyword(items.pop().unwrap()));
+    Tree::constructor_definition(private, name, inline_args, block_args)
+}
+
+fn into_private_keyword(item: Item) -> token::PrivateKeyword {
+    let Item::Token(keyword) = item else { unreachable!() };
+    let token::Variant::PrivateKeyword(variant) = keyword.variant else { unreachable!() };
+    keyword.with_variant(variant)
```
Comment (farmaazon) [inline lib/rust/parser/src/syntax/statement/function_def.rs]: The same function is in `statement.rs`

File: lib/rust/parser/src/syntax/statement/type_def.rs [hunk 4]
```diff
@@ -120,20 +120,24 @@ fn parse_type_body_statement<'s>(
...
+}
+
+fn apply_excess_private_keywords<'s>(
+    mut statement: Option<Tree<'s>>,
+    keywords: impl Iterator<Item = Item<'s>>,
+) -> Option<Tree<'s>> {
+    for token in keywords {
+        let Item::Token(keyword) = token else { unreachable!() };
+        let token::Variant::PrivateKeyword(variant) = keyword.variant else { unreachable!() };
         let keyword = keyword.with_variant(variant);
```
Comment (farmaazon) [inline lib/rust/parser/src/syntax/statement/type_def.rs]: Another `into_private_keyword` logic

File: lib/rust/parser/src/syntax/tree.rs [hunk 5]
```diff
@@ -104,10 +104,9 @@ macro_rules! with_ast_definition { ($f:ident ($($args:tt)*)) => { $f! { $($args)
         Ident {
             pub token: token::Ident<'s>,
         },
-        /// A `private` keyword, marking associated expressions as project-private.
+        /// A `private` declaration.
         Private {
-            pub keyword: token::Private<'s>,
-            pub body: Option<Tree<'s>>,
+            pub keyword: token::PrivateKeyword<'s>,
```
Comment (farmaazon) [inline lib/rust/parser/src/syntax/tree.rs]: This is for making a module private, right? I would then add this info here.

**Review by farmaazon — APPROVED** [review]

File: engine/runtime-integration-tests/src/test/java/org/enso/compiler/test/ErrorCompilerTest.java (hunk 1 above)
Comment (kazcw) [reply engine/runtime-integration-tests/src/test/java/org/enso/compiler/test/ErrorCompilerTest.java:494]: There's no risk of rejecting this due to the one-private-declaration-per-module rule--the second `private` is not a declaration but a field of the `Function`.

File: lib/rust/parser/src/syntax/statement/function_def.rs (hunk 3 above)
Comment (kazcw) [reply lib/rust/parser/src/syntax/statement/function_def.rs]: Implemented some downcasting helpers so we don't need a special function for this case.

**Pull request merged**